### PR TITLE
Add support for generating perf maps for simple perf profiling

### PR DESCRIPTION
**Pull request by bnjbvr**
This adds support for a simpler way to profile with `perf`, using perf maps files. While `jitdump` support is pretty effective and complete when you have access to DWARF metadata, the `perf-inject` step may take lots of time, and it is not optimized at all when modules contain lots of functions. (In our embedding, I've stopped `perf-inject` after it ran for 5 hours and didn't complete, while creating 175K `.so` objects.) Simplistic support for perf maps enable basic profiling: caller/callee trees, top/bottom time analysis, flamegraphs etc., and works very nicely when the wasm modules have many functions.

I've also tweaked the documentation and tried to make the feature available in the C API too.

I haven't put this behind a feature toggle because this profiling agent does very little, so it's always compiled _when running Linux_. For consistency I could add one, if you prefer.

### CODE REVIEW

File: crates/cli-flags/src/lib.rs [hunk 1]
```diff
@@ -143,11 +149,15 @@ pub struct CommonOptions {
...
 
-    /// Generate vtune (supported on --features=vtune build)
-    #[clap(long, conflicts_with = "jitdump")]
+    /// Generate perf mapping file
+    #[clap(long, conflicts_with_all = &["vtune", "jitdump"])]
+    pub perfmap: bool,
+
+    /// Generate vtune runtime information (supported on --features=vtune build)
+    #[clap(long, conflicts_with_all = &["jitdump", "perfmap"])]
     pub vtune: bool,
```
Comment (alexcrichton) [inline crates/cli-flags/src/lib.rs:161]: We may be getting somewhat close to the limit of "now it's probably better to have `--profile=jitdump`" as opposed to a separate flag-per-strategy, but that's fine to defer to a different PR.

**Review by alexcrichton — COMMENTED** [review]
Whoa I had no idea this was a supported format by `perf`! Is there some online documentation for this? Or is there a reason to use `jitdump` over this?

I've got some minor comments but otherwise this seems reasonable to me to add.

**Comment by bjorn3** [comment]
> Or is there a reason to use jitdump over this?

Jitdump can handle code memory being reused by multiple functions at different times. In addition it allows perf to show the actual instructions that executed rather than just the function names. It also supports recording unwind tables as opposed to requiring frame pointers for unwinding. And finally it allows recording line tables.

> Is there some online documentation for this?

https://github.com/torvalds/linux/blob/master/tools/perf/Documentation/jit-interface.txt

File: crates/jit/src/profiling/perfmap_linux.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        let mut file = PERFMAP_FILE.lock().unwrap();
+        if file.is_none() {
+            let filename = format!("/tmp/perf-{}.map", process::id());
+            *file = Some(File::create(filename)?);
+        }
+        Ok(PerfMapAgent)
+    }
+
+    fn make_line(name: &str, addr: *const u8, len: usize) -> String {
+        format!("{:#x} {len} {name}\n", addr as usize)
```
[the remark this answers is not in the document]
Comment (bnjbvr) [reply crates/jit/src/profiling/perfmap_linux.rs:26]: Good point; I've sanitized by removing `\n` and `\r`, not sure if there's any other control character I should be worried about?

**Comment by bnjbvr** [comment]
> I think the performance issues of the jitdump support are because we emit a record for each individual function at 

Nice find @bjorn3! Although I'm not deep enough into the jitdump support to know how to fix this properly and quickly, and perfmap support is a great start for our use case.

File: crates/jit/src/profiling/perfmap_linux.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            let len = func.len();
+            let name = super::debug_name(module, idx);
+            let _ = file.write_all(Self::make_line(&name, addr, len).as_bytes());
+        }
+
+        // Note: these are the trampolines into exported functions.
+        for (idx, func, len) in module.trampolines() {
+            let (addr, len) = (func as usize as *const u8, len);
+            let name = format!("wasm::trampoline[{}]", idx.index());
+            let _ = file.write_all(Self::make_line(&name, addr, len).as_bytes());
```
Comment (alexcrichton) [inline crates/jit/src/profiling/perfmap_linux.rs:51]: Ah I missed this earlier, but I think we'll want ot do something other than ignoring errors here ideally. If an error happens it should probably "close" the file and terminate all future writing to it I suspect? (along with perhaps a warning message printed?)

**Review by alexcrichton — APPROVED** [review]
Thanks for the info and the link @bjorn3! I'm not sure what I expected in the way of documentation but... that I suppose at least does what it says on the tin! Makes sense we'll definitely want to keep the jitdump interface for "more robust" usage.

>  instead of emitting a single record covering the entire text section of the module as intended.

Oh I didn't realize that it was possible to do this, but reading over [the documentation](https://raw.githubusercontent.com/torvalds/linux/master/tools/perf/Documentation/jitdump-specification.txt) it looks like `JIT_CODE_LOAD` only supports one symbol? Do you know off the top of your head how the whole text section would be loaded? I'd love to fix this myself to avoid generating thousands of tiny `*.so` files on all `perf inject` calls.

**Comment by bjorn3** [comment]
> Oh I didn't realize that it was possible to do this, but reading over [the documentation](https://raw.githubusercontent.com/torvalds/linux/master/tools/perf/Documentation/jitdump-specification.txt) it looks like JIT_CODE_LOAD only supports one symbol? Do you know off the top of your head how the whole text section would be loaded? I'd love to fix this myself to avoid generating thousands of tiny *.so files on all perf inject calls.

Right, I should have read the docs more carefully. My bad. All we can do I guess is to make `perf inject` not produce a `.so` for each individual function or write some rust code that has the same functionality as `perf inject` but is faster.

**Comment by bjorn3** [comment]
Wow, `perf inject` is dumb. It is not even that we do something wrong causing each function to be a separate elf file. Instead it really produces a elf file for each function unconditionally: https://github.com/torvalds/linux/blob/9c1bec9c0b08abeac72ed6214b723adc224013bf/tools/perf/util/jitdump.c#L466

**Comment by bnjbvr** [comment]
> or write some rust code that has the same functionality as perf inject but is faster.

:eyes: https://github.com/mstange/samply/issues/33

File: crates/jit/src/profiling/perfmap_linux.rs (hunk 3 above)
Comment (bnjbvr) [reply crates/jit/src/profiling/perfmap_linux.rs:51]: Good point! I've added log messages and early-returns. Looks like a `BufWriter` will automatically flush in its Drop impl, so we can avoid doing it explicitly in most early-returns.

**Pull request merged**